docs/source/1.0/features/speculative-decoding.md

```diff
@@ -1 +1,249 @@
+
+Note that there are currently a few limitations on the two model implementation:
+* KV cache reuse must be disabled.
+* Overlap scheduling must be disabled.
+
+In this approach, we introduce two new steps to the `PyExecutor`'s `_executor_loop`.
+* `_prepare_draft_requests`
+* `_prepare_draft_tokens`
+
+### `_prepare_draft_requests`
+
+This stage occurs for all speculative decoding algorithms before scheduling. The purpose
+of this stage is to make the KV cache and scheduler aware of the fact that speculative decoding
+will occur. Draft tokens take up extra KV cache pages and count towards the executor's
+`max_num_tokens` limit. Thus, we need a way to tell the scheduler that drafting will occur
+**before we do the scheduling**.
+
+To achieve this, we simply attach the maximum number of draft tokens to each request. The
+scheduler and KV cache manager will automatically account for tokens attached to the
+`py_draft_tokens` attribute.
+
+```python
+for req in self.active_requests:
+    req.py_draft_tokens = [0] * max_draft_len
+```
+
+### `_prepare_draft_tokens`
+
+This stage occurs after scheduling and KV cache allocation. The purpose of this stage
+is to attach draft tokens to the `py_draft_tokens` attribute. This occurs by calling `self.drafter.prepare_draft_tokens`;
+each speculative decoding algorithm should have a concrete instance of the `Drafter` class associated with it that defines
+the drafting logic.
+
+In addition to producing all "real" draft tokens, `_prepare_draft_tokens` currently must also pad
+all `py_draft_tokens` to the maximum draft length. This is a CUDA graph limitation - the target
+model will capture its CUDA graphs using the maximum number of draft tokens on each request.
+
+### Verification and Sampling
+
+Once the draft tokens are obtained, the target model runs a forward pass through the usual flow.
+Everything is the same, except that the logits for all the draft tokens are returned and passed
+to the sampler.
+
+Currently, only greedy sampling is supported for speculative decoding. A draft token is accepted if
+matches the previously decoded token exactly. For example, suppose we have a generation request
+`[t, d1, d2, d3]`, where `d1`, `d2`, and `d3` are drat tokens. Suppose the token after `t` is `d1`
+(determined with the `argmax` of the logits). `d1` is then accepted. If the token after `d1` is `d2`,
+then we can accept `d2`. And so on until we can't accept draft tokens anymore.
+
+### KV Cache Rewind
+
+KV cache space allocated to rejected tokens is freed before the next iteration. This is achieved by setting
+the `request.py_rewind_len` attribute to `num_draft_tokens_allocated - num_accepted_tokens`. The pages are
+freed as part of the `resource_manager.free_resources` routine.
+
+The purpose of KV cache rewind is to avoid complicated page reuse logic in the KV cache manager's `prepare_resources`
+function. In practice, this is very cheap since we're just marking blocks as available and not actually
```

Comment:
"...since we're just marking blocks as available..." --> "...since the blocks are just marked as available..."